### PR TITLE
Generate new library version

### PR DESCRIPTION
**Pull request by brettmortensen**
First cut at generating a new version of the wrapper library using
code-gen tools.

The generated stock file structure does not look right, I'm thinking of separating the models into a `models` directory and api files into a `client` directory.

Any feedback is appreciated 👍 

@film42 @minond

### CODE REVIEW

File: api_connect_widget.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
+/*
+ * MX API
+ *
+ * The MX Atrium API supports over 48,000 data connections to thousands of financial institutions. It provides secure access to your users' accounts and transactions with industry-leading cleansing, categorization, and classification.  Atrium is designed according to resource-oriented REST architecture and responds with JSON bodies and HTTP response codes.  Use Atrium's development environment, vestibule.mx.com, to quickly get up and running. The development environment limits are 100 users, 25 members per user, and access to the top 15 institutions. Contact MX to purchase production access. 
+ *
+ * API version: 0.1
+ */
+
+package atrium-go
```
Comment (minond) [inline api_connect_widget.go]: This is a syntax error.

File: api_connect_widget.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+			} else {
+				key = auth.Key
+			}
+			localVarHeaderParams["MX-API-Key"] = key
+			
+		}
+	}
+	if ctx != nil {
+		// API Key Authentication
+		if auth, ok := ctx.Value(ContextAPIKey).(APIKey); ok {
```
Comment (minond) [inline api_connect_widget.go:87]: Looks to be using the same key (`ContextAPIKey`) for both the API key and the Client ID values.

File: client.go [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+	return localVarRequest, nil
+}
+
+func (c *APIClient) decode(v interface{}, b []byte, contentType string) (err error) {
+		if strings.Contains(contentType, "application/xml") {
+			if err = xml.Unmarshal(b, v); err != nil {
+				return err
+			}
+			return nil
+		} else if strings.Contains(contentType, "json") {
```
Comment (minond) [inline client.go]: Oh nice, this should take care of the decoding issue.

**Review by minond — CHANGES_REQUESTED** [review]
At first glance things look OK to me. Just the package name issue (can't have a hyphen and it should just be "atrium" IMO) and the ClientID key issue, that's all I noticed. If you need a hand running some of this code LMK.

**Comment by minond** [comment]
@brettmortensen what were the issues we found the first time we tried this? Missing client IDs, weird request/response types, json decoding failing by default, anything else?

File: README.md [hunk 4]
```diff
@@ -1,53 +1,96 @@
...
+# Go API client for atrium-go
 
-This is a wrapper for the [MX Atrium API](https://atrium.mx.com/).
+The MX Atrium API supports over 48,000 data connections to thousands of financial institutions. It provides secure access to your users' accounts and transactions with industry-leading cleansing, categorization, and classification.  Atrium is designed according to resource-oriented REST architecture and responds with JSON bodies and HTTP response codes.  Use Atrium's development environment, vestibule.mx.com, to quickly get up and running. The development environment limits are 100 users, 25 members per user, and access to the top 15 institutions. Contact MX to purchase production access. 
 
-#### Installation
+## Installation
+Put the package under your project folder and add the following in import:
+```golang
+import "./atrium-go"
```
Comment (film42) [inline README.md]: This is wrong. You'll be importing `github.com/mxenabled/atrium-go`.

File: LICENSE [hunk 5]
```diff
@@ -1,21 +0,0 @@
-The MIT License (MIT)
```
Comment (film42) [inline LICENSE]: Why are we removing the MIT license? This code is open source and free to use/ change.

**Comment by brettmortensen** [comment]
@minond I believe you listed all of the issues we found above. I think they have all been resolved now besides the generic request/response types.

**Comment by minond** [comment]
@brettmortensen just messaged you some solutions for the type issues we are running into. Let me know if they help.

**Comment by brettmortensen** [comment]
@minond Thanks, I've updated the library with the solution you suggested. It appears to solve the issue we were seeing with the interfaces. Let me know what you think 👍

**Comment by minond** [comment]
@brettmortensen that's awesome. I took a quick peek at the user create payload body and it looks like it's still using an `interface{}`. Did all of the definitions get updated or just some of them?

**Comment by minond** [comment]
I take back my last comment. I'll take a look later today when I free up a bit more!

**Comment by minond** [comment]
Looking good to me. I like the new constructors you created, really cleans things up. Digging around and the types look much cleaner as well -- happy not to see any `interface{}`'s. The only thing I don't like is the whole "attributes" convention that we have right now. For example, `User` is defined as

```go
type User struct {
	User *UserAttributes `json:"user,omitempty"`
}
```

so creating one requires doing

```go
user := atrium.User{
	User: &atrium.UserAttributes{
		Guid: "123",
	},
}
```

but really I just want to be able to

```go
user := atrium.User{
	Guid: "123",
}
```

I'm not sure if there's a reason we're having to do this right now, but if we could do away with a "ModelAttribute" and just work with a "Model" that'd be nice. I think we should clean this up before releasing this.

The other thing is more a nitpick than anything else, and that's that we should follow Go's naming convention for things like "URL", "GUID", "ID", etc. Running a linter on this project (https://github.com/golang/lint) I'm seeing lots of warnings about "userGuid" needing to be "userGUID", "MediumLogoUrl" needing to be "MediumLogoURL", and so on. If you do a `go get -u golang.org/x/lint/golint` and then run `golint` at the root of this project you can see the full list yourself. You'll see warnings about missing comments, using underscores, and a few other naming conventions not being followed. I say we focus on properly naming model fields (eg `User.Guid -> User.GUID`) and API objects (eg `client.UsersApi -> client.UsersAPI`) right now and worry less about the rest for now. I think those names are more important since users will actually be using those.

**Comment by brettmortensen** [comment]
@minond Thanks, great feedback. 

The naming of the model fields and API objects has been updated.

Good point about the "attributes" pattern I followed. When setting this up originally, I followed some swagger examples that used the same pattern. I've been searching/trying to get rid of the nested "attributes" model, but it doesn't appear there is an easy way to do that in the generators. Any time there are nested objects, it forces them to be individual objects 👎

**Comment by brettmortensen** [comment]
@minond I just updated the models with the "request"/"response" pattern we discussed yesterday offline. It should clear up the confusion with the previous "attributes" pattern.

**Comment by film42** [comment]
:shipit: 

Let's make sure to tag this as `v2.0.0`.

**Pull request merged**